packages/remix/src/migrations/update-14-5-4/update-tsconfig-and-remix-config-for-1-6-8.ts

```diff
@@ -0,0 +1,102 @@
+import {
+  formatFiles,
+  getProjects,
+  getWorkspaceLayout,
+  joinPathFragments,
+  logger,
+  offsetFromRoot,
+  readJson,
+  Tree,
+  writeJson,
+} from '@nrwl/devkit';
+
+/**
+ * Update tsconfig.json and remix.config.js to support importing workspaces libraries
+ * @param tree
+ */
+export default async function update(tree: Tree) {
+  const projects = getProjects(tree);
+
+  const remixProjects = Array.from(projects.values()).filter((project) => {
+    const remixConfigPath = joinPathFragments(project.root, 'remix.config.js');
+    // if the project doesn't have a remix.config.js, it's not a Remix project so we can skip it
+    return tree.exists(remixConfigPath);
+  });
+
+  const remixProjectsWithDefaultPaths = remixProjects.filter((project) => {
+    const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
+    const tsConfigContent = readJson(tree, tsConfigPath);
+
+    return !!tsConfigContent.compilerOptions.paths['~/*'];
+  });
+
+  const updateTsConfigPaths = remixProjectsWithDefaultPaths.length === 1;
+
+  if (remixProjects.length === 0) return;
+
+  remixProjects.forEach((project) => {
+    const remixConfigPath = joinPathFragments(project.root, 'remix.config.js');
+
+    try {
+      const remixConfigContent = tree.read(remixConfigPath, 'utf-8');
+      // if watchPaths is already there, we assume this project has been updated before
+      if (remixConfigContent.includes('watchPaths')) return;
+
+      const projectOffsetFromRoot = offsetFromRoot(project.root);
+      const layout = getWorkspaceLayout(tree);
+      tree.write(
+        remixConfigPath,
+        remixConfigContent.replace(
+          'module.exports = {\n',
+          `module.exports = {  watchPaths: ['${joinPathFragments(
+            projectOffsetFromRoot,
+            layout.libsDir
+          )}'],\n`
+        )
+      );
+
+      const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
+      const tsConfigContent = readJson(tree, tsConfigPath);
+
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const newTsConfigContent = {
+        ...tsConfigContent,
+        extends: `${projectOffsetFromRoot}tsconfig.base.json`,
+      };
+
+      delete newTsConfigContent.compilerOptions.baseUrl;
+
+      if (updateTsConfigPaths) {
+        delete newTsConfigContent.compilerOptions.paths;
+
+        const tsconfigBaseJsonPath = 'tsconfig.base.json';
+        const tsConfigBaseJsonContent = readJson(tree, tsconfigBaseJsonPath);
+
+        const newTsConfigBaseJsonContent = {
+          ...tsConfigBaseJsonContent,
+          compilerOptions: {
+            ...tsConfigBaseJsonContent.compilerOptions,
+            paths: {
+              ...tsConfigBaseJsonContent.compilerOptions.paths,
+              '~/*': [joinPathFragments(project.root, 'app', '*')],
```

Comment:
I see what you did here... but should we 🤔. What if they add another app and intuitively use `~/.....` in their imports. It would then import stuff from the 1st app that thas been there.

Unfortunately, you cannot merge TS path mappings (which is also understandable)

Reply:
That's a good point. 🤔 I don't personally like the `~` path. In an Nx workspace, the majority of your code should be down in libraries, so you don't have to worry too much about deeply nested relative paths in your app. But I think it's an accepted standard in the Remix community. Some alternatives:

* Don't migrate anything automatically, add a prominent log during migration saying to handle the migration yourself. This is already done if this migration detects multiple Remix apps. Users could probably do this pretty quickly with a search and replace.
* Provide a generator that would allow the user to define a new path for that app, maybe something like `~my-app`. The generator would update the `tsconfig` files and change all of the import paths. We would probably want to use the TypeScript AST to be safe with this update, but I don't think it would be too hard.

Reply:
> But I think it's an accepted standard in the Remix community

The issue I'm mostly having is that it wouldn't make a lot of sense in a monorepo scenario. If we look at it from using Nx for getting quickly set up with Remix etc, so "Nx as a Remix CLI", then we can totally leave it. But from a monorepo perspective it doesn't make sense to have it in there.

Whether we do the migration or not really depends on how much effort we want to put into this, given the Remix plugin is still a labs plugin. I'd say to not migrate and rather have the log for now.

Reply:
I've removed the migration and added a log to the migration script.